src/cpu/flexcpu/simple_dataflow_cpu.hh

```diff
@@ -0,0 +1,496 @@
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#ifndef __CPU_FLEXCPU_SIMPLE_DATAFLOW_CPU_HH__
+#define __CPU_FLEXCPU_SIMPLE_DATAFLOW_CPU_HH__
+
+#include <functional>
+#include <list>
+#include <memory>
+#include <vector>
+
+#include "cpu/base.hh"
+#include "cpu/flexcpu/simple_dataflow_thread.hh"
+#include "mem/packet.hh"
+#include "params/SimpleDataflowCPU.hh"
+
+class SDCPUThread;
+/**
+ * The CPU class represents a core, which contains hardware capable of running
+ * one or more threads. This class has the responsibility of managing hardware
+ * structures and controlling the timing of events.
+ */
+class SimpleDataflowCPU : public BaseCPU
+{
+  public:
+    using ExecCallback = std::function<void(Fault)>;
+    using FetchCallback = std::function<void(TheISA::MachInst)>;
+    using MemCallback = std::function<void(Fault)>;
+    // May need to also return an ID of some sort through the MemCallback,
+    // since we may need a way for threads to request the squashing of a
+    // request that hasn't actually gone to memory yet.
+    using TranslationCallback = std::function<void(Fault,const RequestPtr&)>;
+  protected:
+    // BEGIN Internal class definitions
+
+    class DataPort : public MasterPort
+    {
+      protected:
+        SimpleDataflowCPU* cpu;
+      public:
+        bool receivedPushback = false;
+
+        DataPort(const std::string& name_, SimpleDataflowCPU* cpu_):
+            MasterPort(name_, cpu_),
+            cpu(cpu_)
+        { }
+
+        virtual ~DataPort()
+        { }
+
+        bool recvTimingResp(PacketPtr pkt);
+
+        void recvReqRetry();
+    };
+
+    class InstPort : public MasterPort
+    {
+      protected:
+        SimpleDataflowCPU* cpu;
+      public:
+        bool receivedPushback = false;
+
+        InstPort(const std::string& name_, SimpleDataflowCPU* cpu_):
+            MasterPort(name_, cpu_),
+            cpu(cpu_)
+        { }
+
+        virtual ~InstPort()
```

Comment:
If we're talking about the destructor, I often default to declaring a virtual destructor when defining classes, simply because I rarely feel like there's a reason for a destructor not to be virtual. Maybe it's because I'm used to Java where there aren't restrictions on what classes you can subclass.